### PR TITLE
Support safe call

### LINKED ISSUE #40: Safe navigation operator is not supported
I would like to implement that but I am not touching the codebase until someone takes a look on my previous 5 PRs, so leaving this as an issue:

```
     20:   def test_safe_navigation_operator
  => 21:     assert @box.eval('x = 1; x&.to_s')
     22:   end
     23:
     24:   def test_unary_operators_on_instances_of_boolean_vars
/home/lzap/work/safemode/lib/safemode.rb:49:in `eval'
/home/lzap/work/safemode/lib/safemode/parser.rb:10:in `jail'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/sexp_processor-4.11.0/lib/sexp_processor.rb:245:in `process'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/sexp_processor-4.11.0/lib/sexp_processor.rb:388:in `in_context'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/sexp_processor-4.11.0/lib/sexp_processor.rb:268:in `block in process'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/sexp_processor-4.11.0/lib/sexp_processor.rb:324:in `error_handler'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/sexp_processor-4.11.0/lib/sexp_processor.rb:269:in `block (2 levels) in process'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby2ruby-2.4.1/lib/ruby2ruby.rb:195:in `process_block'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby2ruby-2.4.1/lib/ruby2ruby.rb:195:in `map'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby2ruby-2.4.1/lib/ruby2ruby.rb:196:in `block in process_block'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/sexp_processor-4.11.0/lib/sexp_processor.rb:245:in `process'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/sexp_processor-4.11.0/lib/sexp_processor.rb:388:in `in_context'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/sexp_processor-4.11.0/lib/sexp_processor.rb:268:in `block in process'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/sexp_processor-4.11.0/lib/sexp_processor.rb:324:in `error_handler'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/sexp_processor-4.11.0/lib/sexp_processor.rb:269:in `block (2 levels) in process'
/home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby2ruby-2.4.1/lib/ruby2ruby.rb:297:in `process_safe_call'
/home/lzap/work/safemode/lib/safemode/parser.rb:35:in `process_call'
Error: test_safe_navigation_operator(TestSafemodeEval): ArgumentError: wrong number of arguments (given 2, expected 1)
```

### PR DESCRIPTION
**Pull request by ekohl**
This copies the process_call_code method from ruby2ruby. That also implements process_safe_call which calls process_call with a parameter.

Fixes #40

### CODE REVIEW

File: lib/safemode/parser.rb [hunk 1]
```diff
@@ -159,25 +162,39 @@ def process_call_args(exp)
...
         args << processed unless (processed.nil? or processed.empty?)
       end
-      args.empty? ? nil : args.join(", ")
+      args
     end
 
-    def process_call_code(receiver, name, args)
+    def process_call_code(receiver, name, args, safe_call)
       case name
-      when :<=>, :==, "!=".to_sym, :<, :>, :<=, :>=, :-, :+, :*, :/, :%, :<<, :>>, :** then
```
Comment (ekohl) [inline lib/safemode/parser.rb:167]: Looking at https://github.com/seattlerb/ruby2ruby/blob/ddf539eaf348b57c35acd91109fc6d368289b517/lib/ruby2ruby.rb#L39-L40 this implies we'll also support `|` and `&` operators which previously weren't supported. I'm not sure if there's a security risk involved there.

**Comment by ekohl** [comment]
Trivial rebase.

File: lib/safemode/parser.rb [hunk 2]
```diff
@@ -159,25 +162,39 @@ def process_call_args(exp)
         end
         args << processed unless (processed.nil? or processed.empty?)
       end
-      args.empty? ? nil : args.join(", ")
```
Comment (ofedoren) [inline lib/safemode/parser.rb:162]: Not quite familiar with what's that for, but why is this change?

File: lib/safemode/parser.rb (hunk 1 above)
Comment (ofedoren) [reply lib/safemode/parser.rb:167]: For the sake of completeness, we will support `^` as well. How would someone misuse these operators?.. @adamruzicka, maybe you have an idea? :)

**Review by ofedoren — COMMENTED** [review]
Thanks, @ekohl, some comments inline:

File: lib/safemode/parser.rb (hunk 1 above)
Comment (adamruzicka) [reply lib/safemode/parser.rb:167]: Looking at https://github.com/theforeman/safemode/blob/01af658932811800e2fc0542db4332a922192131/lib/safemode/core_jails.rb#L30 I'd say we did support them even before, but in a non-operator kind of way?

Before this change when I try to render `<%= 5 ^ 3 %>`
```
> receiver
=> "5.to_jail"
> name
=> :^
> args
=> "3"
> "#{receiver}.#{name}#{args ? "(#{args})" : args}"
=> "5.to_jail.^(3)"
```

With this change the generated code will be `5.to_jail ^ 3`, but semantically it should be the same thing so we don't really need to worry about it?

File: lib/safemode/parser.rb [hunk 3]
```diff
@@ -26,19 +26,22 @@ def parser=(parser)
...
+      str = if str
+              dot = safe_call ? "&." : "."
+              parentheses ? "(#{str})&#{dot}" : "#{str}#{dot}"
+            end
       "#{str}to_jail"
     end
 
     # split up #process_call. see below ...
-    def process_call(exp)
+    def process_call(exp, safe_call = false)
```
[the remark this answers is not in the document]
Comment (adamruzicka) [reply lib/safemode/parser.rb:38]: If I'm reading things right then
- ruby_parser generates a :safe_call sexp when it encounters `&.` https://github.com/seattlerb/ruby_parser/blob/8f419fdbe7e58bf039cd948b426b18c2f9158548/lib/ruby_parser_extras.rb#L882
- sexp processor does [some magic](https://github.com/seattlerb/sexp_processor/blob/master/lib/sexp_processor.rb#L153) to generate various `process_*` methods, these methods get called as it walks the tree of sexps ([somewhere around here](https://github.com/seattlerb/sexp_processor/blob/master/lib/sexp_processor.rb#L260))
- ruby2ruby has `process_safe_call(exp)`, which just calls `process_call(exp, :safe)`

**Review by ofedoren — APPROVED** [review]
Thanks, @ekohl and @adamruzicka, I think this is ready.

**Pull request merged**